### PR TITLE
improve test scripts

### PR DESCRIPTION
**Pull request by zimbatm**
* validated with shellcheck
* allow to evaluate just one for convenience
* reduce verbosity and improve debugging output

### CODE REVIEW

File: tests/eval-one.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#!/bin/sh
```
Comment (yegortimoshenko) [inline tests/eval-one.sh]: Evaluating just one is not a proper test, because profiles can be inherited. It can be done via `nixos-rebuild -I nixos-config=tests/eval-test.nix nixos-hardware-config=path/to/default.nix dry-build`. I think it's a bad idea to allow to eval just one via a script.

File: tests/eval-test.sh [hunk 2]
```diff
@@ -1,15 +1,11 @@
 #!/bin/sh
+set -efu
+cd "$(dirname "$0")/.." || exit 1
```
Comment (yegortimoshenko) [inline tests/eval-test.sh]: This can't really fail.

File: tests/eval-one.sh [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  -I "nixos-config=$here/eval-test.nix" \
+  -I "nixos-hardware-profile=$profile" \
+  --dry-run \
+  --no-out-link \
+  --show-trace \
+  -A system > "$logfile" 2>&1
+ret=$?
+
+if [ "$ret" -gt 0 ]; then
+  echo ERROR
```
Comment (yegortimoshenko) [inline tests/eval-one.sh]: `nix-build` already prints error, we don't need to print that twice.

File: tests/eval-one.sh [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+# shellcheck disable=SC2039
+echo -n "evaluating $profile: "
+
+nix-build \
+  '<nixpkgs/nixos>' \
+  -I "nixos-config=$here/eval-test.nix" \
+  -I "nixos-hardware-profile=$profile" \
+  --dry-run \
+  --no-out-link \
```
Comment (yegortimoshenko) [inline tests/eval-one.sh]: `--no-out-link` doesn't do anything when `--dry-run` is present.

File: tests/eval-one.sh [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+profile=$1
+logfile=$(mktemp)
+
+# shellcheck disable=SC2039
+echo -n "evaluating $profile: "
+
+nix-build \
+  '<nixpkgs/nixos>' \
+  -I "nixos-config=$here/eval-test.nix" \
+  -I "nixos-hardware-profile=$profile" \
```
Comment (yegortimoshenko) [inline tests/eval-one.sh]: We shouldn't allow profiles that have spaces in names.

File: tests/eval-one.sh [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  --dry-run \
+  --no-out-link \
+  --show-trace \
+  -A system > "$logfile" 2>&1
```
Comment (yegortimoshenko) [inline tests/eval-one.sh]: I don't see the point of having a log file instead of stderr, and I think it hides a lot of potentially useful information. Also, the idea of stderr is exactly to have an error log.

File: tests/eval-test.sh [hunk 7]
```diff
@@ -1,15 +1,11 @@
...
 
-set -e
-
-cd "$(dirname "$0")/.."
+echo "### Evaluating all profiles ###"
```
Comment (yegortimoshenko) [inline tests/eval-test.sh]: This is redundant.

File: tests/eval-one.sh [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+if [ -z "$1" ]; then
+  echo "Usage: $0 <profile>"
+  exit 1
+fi
+
+profile=$1
+logfile=$(mktemp)
+
+# shellcheck disable=SC2039
```
Comment (yegortimoshenko) [inline tests/eval-one.sh]: What's the point of using a tool when it requires magic comments to do its job? I think `shellcheck` is very overreaching, and we shouldn't use it.

File: tests/eval-one.sh [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+echo -n "evaluating $profile: "
+
+nix-build \
+  '<nixpkgs/nixos>' \
+  -I "nixos-config=$here/eval-test.nix" \
+  -I "nixos-hardware-profile=$profile" \
+  --dry-run \
+  --no-out-link \
+  --show-trace \
```
Comment (yegortimoshenko) [inline tests/eval-one.sh]: `--show-trace` might be useful and should make it into `master`.

File: tests/eval-test.sh [hunk 10]
```diff
@@ -1,15 +1,11 @@
...
+echo
 
+# shellcheck disable=SC2044
```
Comment (yegortimoshenko) [inline tests/eval-test.sh]: Not sure why you ignore this check: https://github.com/koalaman/shellcheck/wiki/SC2044

In this case, it could be rewritten as `find . -name default.nix -exec tests/eval-one.sh {}`. But again, I think we shouldn't accept files with weird names into this repo, and current shell script makes sure that we won't.

File: tests/eval-one.sh (hunk 6 above)
Comment (Mic92) [reply tests/eval-one.sh]: No, in the error case the log file contains both stderr and stdin.

File: tests/eval-test.sh [hunk 11]
```diff
@@ -1,15 +1,11 @@
 #!/bin/sh
+set -efu
```
Comment (Mic92) [inline tests/eval-test.sh]: `set -o pipefail` also comes handy.

Comment (yegortimoshenko) [reply tests/eval-test.sh]: It is a bashism.

File: tests/eval-test.sh (hunk 2 above)
Comment (Mic92) [reply tests/eval-test.sh]: `exit 1` should not be necessary, since `set -e` is used.

File: tests/eval-one.sh (hunk 6 above)
Comment (yegortimoshenko) [reply tests/eval-one.sh]: It doesn't matter, I think `nix-build` only ever writes to `stderr`. What matters is that CI builds may stutter, and with this approach, we would never know why.

File: tests/eval-one.sh (hunk 1 above)
Comment (Mic92) [reply tests/eval-one.sh]: Travis will evaluate all of them in the end, or not?

Comment (yegortimoshenko) [reply tests/eval-one.sh]: Travis evaluates every profile on each commit/PR.

File: tests/eval-one.sh (hunk 6 above)
Comment (yegortimoshenko) [reply tests/eval-one.sh]: Also `2>&1` is a bashism.

File: tests/eval-one.sh (hunk 5 above)
Comment (zimbatm) [reply tests/eval-one.sh]: It doesn't hurt to escape the argument in either case

File: tests/eval-one.sh (hunk 8 above)
Comment (zimbatm) [reply tests/eval-one.sh]: It catches a lot of common mistakes and educates the users about shell corner-cases.

File: tests/eval-one.sh (hunk 1 above)
Comment (zimbatm) [reply tests/eval-one.sh]: The idea here is that I can test my profile without having to wait on all the others to evaluate. Travis is testing all the profiles anyways.

File: tests/eval-one.sh (hunk 6 above)
Comment (zimbatm) [reply tests/eval-one.sh]: @yegortimoshenko With this approach it's easy to find the failing profile and get to the point instead of going through pages of repeating out links. They might be interesting but not very useful in practice.

File: tests/eval-test.sh (hunk 10 above)
Comment (zimbatm) [reply tests/eval-test.sh]: There is no solution that doesn't depend on word splitting, without depending on bash 4.  `find -exec` doesn't early-return nor propagate the exit status. I don't think that profiles should contain special characters in any case.

File: tests/eval-one.sh (hunk 3 above)
Comment (zimbatm) [reply tests/eval-one.sh]: Look at the Travis CI output logs and you will understand.

Comment (yegortimoshenko) [reply tests/eval-one.sh]: I don't understand: https://travis-ci.org/NixOS/nixos-hardware/builds/321451543#L5488

File: tests/eval-one.sh (hunk 8 above)
Comment (yegortimoshenko) [reply tests/eval-one.sh]: I don't think we need that for fifteen lines of shell script.

File: tests/eval-one.sh (hunk 1 above)
Comment (yegortimoshenko) [reply tests/eval-one.sh]: You shouldn't assume that your profile isn't inherited by any other profiles. Profile might be inherited by someone from another directory, e.g. in case of brand renaming (e.g. IBM -> Lenovo).

File: tests/eval-one.sh (hunk 5 above)
Comment (yegortimoshenko) [reply tests/eval-one.sh]: It does. We should enforce that there is no white spaces in names.

File: tests/eval-one.sh (hunk 6 above)
Comment (yegortimoshenko) [reply tests/eval-one.sh]: Right, you only need logs if something goes wrong. Your way of doing things makes certain errors opaque and non-debuggable.

File: tests/eval-one.sh (hunk 1 above)
Comment (Mic92) [reply tests/eval-one.sh]: How is this relevant? Travis will catch this eventually.

Comment (yegortimoshenko) [reply tests/eval-one.sh]: It's relevant in that it's not a test and thus should not be exposed. It's possible to see someone with merge rights using `eval-one` test and then pushing to `master`.

**Comment by yegortimoshenko** [comment]
Resolved in #66 with a fundamentally different approach.

**Comment by yegortimoshenko** [comment]
@zimbatm I shouldn't have behaved so aggressively over this, I'm sorry.
